plugins/woocommerce/client/blocks/assets/js/extensions/payment-methods/paypal/buttons.js

```diff
@@ -0,0 +1,188 @@
+/**
+ * External dependencies
+ */
+import { useState } from '@wordpress/element';
+import { PayPalScriptProvider, PayPalButtons } from '@paypal/react-paypal-js';
+import { getPaymentMethodData } from '@woocommerce/settings';
+import { dispatch } from '@wordpress/data';
+import apiFetch from '@wordpress/api-fetch';
+
+/**
+ * PayPalButtonsContainer component.
+ *
+ * @param {Object}  props
+ * @param {string}  props.clientId
+ * @param {string}  [props.components]
+ * @param {string}  [props.disableFunding]
+ * @param {string}  [props.enableFunding]
+ * @param {string}  [props.currency]
+ * @param {string}  [props.intent]
+ * @param {string}  [props.merchantId]
+ * @param {string}  [props.partnerAttributionId]
+ * @param {string}  [props.pageType]
+ * @param {boolean} [props.isProductPage]
+ * @return {JSX.Element} The PayPal Buttons container component.
+ */
+const PayPalButtonsContainer = ( {
+	clientId,
+	components,
+	disableFunding,
+	enableFunding,
+	currency,
+	intent,
+	merchantId,
+	partnerAttributionId,
+	pageType,
+	isProductPage,
+} ) => {
+	const [ orderReceivedUrl, setOrderReceivedURL ] = useState( '' );
+	const [ orderId, setOrderId ] = useState( '' );
+	const payPalData = getPaymentMethodData( 'paypal', {} );
+	const options = {
+		clientId: clientId || '',
+		components: components || '',
+		disableFunding: disableFunding || '',
+		enableFunding: enableFunding || '',
+		currency: currency || '',
+		intent: intent || '',
+		merchantId: merchantId || '',
+		'data-partner-attribution-id': partnerAttributionId || '',
+		'data-page-type': pageType || '',
+	};
+
+	const createOrder = async ( data ) => {
+		let responseData;
+		try {
+			// If we're inside the product page, we need to empty the cart,
+			// and add the current product to the cart.
+			if ( isProductPage ) {
+				// Empty the cart.
+				await apiFetch( {
+					method: 'DELETE',
+					path: '/wc/store/v1/cart/items',
+				} );
+
+				// Get product ID from the value of the "add-to-cart" button.
+				let productId = document.querySelector(
+					'[name="add-to-cart"]'
```

Comment:
These selector seem a little brittle.  I think ideally these would be passed via hydration from the server so we're not risking a selector that could be overwritten by the theme.

There are many ways and this might not be the best example, but here is one way that context is passed via the server: https://github.com/woocommerce/woocommerce/blob/77b902e5f8ee6fadbf798296fe4c3bfc52b67165/plugins/woocommerce/src/Blocks/BlockTypes/ProductGallery.php#L140

Reply:
Valid concern. I do think though that there will always be a form field named 'add-to-cart', as this is what [the form submission action relies on](https://github.com/woocommerce/woocommerce/blob/0dd4b8efc2c44cb9c7bc6a5c7e8324cb52390ee8/plugins/woocommerce/includes/class-wc-form-handler.php#L823).

We can pass the product ID from the server as suggested, but we will still have to rely on form fields for quantity (and variation ID for variable products). LMK if there is a better way for quantity and variation ID.

I think we can keep this as-is for now. We are doing the same for Stripe's express checkout, and AFAIK no issues around product IDs reported yet.